test/cql-pytest/test_validation.py

```diff
@@ -0,0 +1,368 @@
+    for b in good_utf8:
+        print(b)
+        cql.execute(cmd.format(table1, b.hex()))
+        # verify that the successfully-written value can be read correctly
+        results = list(cql.execute(f"SELECT k, t FROM {table1} WHERE k=1"))
+        assert len(results) == 1
+        assert results[0].k == 1 and results[0].t.encode('utf-8') == b
+    for b in bad_utf8:
+        print(b)
+        with pytest.raises(InvalidRequest, match='not a valid binary representation for type text'):
+            cql.execute(cmd.format(table1, b.hex()))
+
+# 2. We can pass the string as a bound argument to a prepared  statement.
+#    Convincing Python to put an invalid UTF-8 here is not trivial, because
+#    the driver outputs strings, which are not supposed to be able to contain
+#    invalid UTF-8. We use a rather funky workaround here use a wrapped
+#    version ("surrogateescape") of bad UTF-8, and monkey-patch the driver to
+#    unescape it when converting it back to bytes.
+# Note that currently, Scylla's UTF-8 parser is stricter than Cassandra's
+# (see comment above listing the relevant cases), so this test, as all tests
+# using the bad_utf8 array, will fail on Cassandra.
+def test_validation_utf8_bound_column(cql, table1):
+    import cassandra.cqltypes
+    orig_serialize = cassandra.cqltypes.UTF8Type.serialize
+    def myserialize(ustr, protocol_version):
+        return ustr.encode('utf-8', errors='surrogateescape')
+    cassandra.cqltypes.UTF8Type.serialize = myserialize
+    try:
+        stmt = cql.prepare(f'INSERT INTO {table1} (k, t) VALUES (1, ?)')
+        for b in good_utf8:
+            print(b)
+            cql.execute(stmt, [b.decode()])
+            results = list(cql.execute(f"SELECT k, t FROM {table1} WHERE k=1"))
+            assert len(results) == 1
+            assert results[0].k == 1 and results[0].t.encode('utf-8') == b
+        for b in bad_utf8:
+            print(b)
+            # Scylla prints "Exception while binding column t: marshaling error:
+            # Validation failed - non-UTF8 character in a UTF8 string, at byte
+            # offset 0". Cassandra prints "String didn't validate.". The only
+            # thing in common is the word 'validat' :-)
+            with pytest.raises(InvalidRequest, match=re.compile('validat', re.IGNORECASE)):
+                cql.execute(stmt, [b.decode(errors='surrogateescape')])
+    finally:
+        cassandra.cqltypes.UTF8Type.serialize = orig_serialize
+
+# 3. We can also insert the bad UTF-8 as part of the request string itself.
+#    This will make the entire request string invalid UTF-8, not just the
+#    value to be inserted, so Scylla and Cassandra should complain that the
+#    entire request is bad - not just the inserted value.
+# FIXME: this test is INCOMPLETE! It's very hard to get Python to output an
+# illegal UTF-8 string in this case, and I gave up. Unlike the prepared-
+# statement case above where it was easy to find and monkey-patch the
```

Comment:
This monkey-patch sounds like a time bomb, exploding when the driver implementation changes.

Reply:
This is true, but it's not really a dangerous bomb - perhaps a stink bomb :-) the most likely consequence of such a driver change would be that the test will fail, and at that point we can either quickly fix the monkey-patching or mark the test "skip".